pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,139 @@
+package vitessshard
+
+import (
+	"context"
+	"sort"
+
+	"k8s.io/api/core/v1"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+	"planetscale.dev/vitess-operator/pkg/operator/vttablet"
+)
+
+func (r *ReconcileVitessShard) reconcileRollout(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
+	if !rollout.Cascaded(vts) {
+		// If the shard is not scheduled for a cascading update, silently bail out and do nothing.
+		return resultBuilder.Result()
+	}
+
+	podList := &v1.PodList{}
+	listOpts := &client.ListOptions{
+		Namespace: vts.Namespace,
+		LabelSelector: apilabels.Set(map[string]string{
+			planetscalev2.ComponentLabel: planetscalev2.VttabletComponentName,
+			planetscalev2.ClusterLabel:   vts.Labels[planetscalev2.ClusterLabel],
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	// Safety checks and rolling updates must be deterministically ordered, so we access and sort pods by tablet alias.
+	tabletPods := make(map[string]*v1.Pod)
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		tabletAlias := vttablet.AliasFromPod(pod)
+		tabletKey := topoproto.TabletAliasString(&tabletAlias)
+		tabletPods[tabletKey] = pod
+	}
+
+	tabletKeys := make([]string, 0, len(tabletPods))
+	for key := range tabletPods {
+		tabletKeys = append(tabletKeys, key)
+	}
+	sort.Strings(tabletKeys)
+
+	for _, tabletKey := range tabletKeys {
+		tablet := vts.Status.Tablets[tabletKey]
+		if tablet.Available != v1.ConditionTrue {
+			// If all tablets aren't healthy, we should bail and not perform a rolling restart.
```

Comment:
Just to be a little clearer:

```suggestion
			// If any tablets are unhealthy, we should bail and not perform a rolling restart.
```